sysvabi64/sysvabi64.rst

```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+-------------------------------------------
+
+The figure below depicts the fundamental components of the TLS
+addressing architecture used by SystemV for AArch64.
+
+.. _SystemV AArch64 TLS addressing architecture:
+
+.. figure:: sysvabi64-tls.svg
+
+   SystemV AArch64 TLS addressing architecture
+
+The TLS data for a module is called the TLS Block.
+
+The thread pointer points directly to the Thread Control Block (TCB).
+
+The size of the TCB is 16-bytes, where the first 8 bytes contain the
+pointer to the Dynamic Thread Vector (DTV), and the other 8 bytes are
+reserved for the implementation.
+
+Following the TCB and any required alignment padding, the TLS Blocks
+of the modules loaded at process start form the static TLS Block. The
+memory for the TLS Block is allocated at process start time.
+
+The TLS Blocks for modules loaded dynamically via dlopen are known as
+dynamic TLS.
+
+Index 0 of the Dynamic Thread Vector DTV[0] typically contains a
+generation counter which can be used to update or reallocate the DTV
+when dynamic modules are opened or closed.
+
+Index N, where N > 0, of the Dynamic Thread Vector DTV[N] points to
+the TLS block for module N.
+
+To calculate the address of a TLS variable in any given module, static
+or dynamic, the expression ``TP[0][Module id][offset in module]`` can
+be used. The function ``__tls_get_address(module_id, offset)`` returns
+the result of this calculation.
+
+The calculation above is the most general and it can be applied to both static and dynamic TLS. There are four defined models of accessing TLS that trade off generality for performance. In order of descending generality:
+
+   1. General Dynamic, can be used anywhere.
+
+   2. Local Dynamic, can be used anywhere where the definition of the
+      TLS variable and the access are from the same module.
+
+   3. Initial Exec, can be used for TLS variables defined in the
+      static TLS block.
+
+   4. Local Exec, can be used in the executable for TLS variables
+      defined in the executables static TLS block.
+
+SystemV AArch64 TLS addressing
+------------------------------
+
+AArch64 TLS SystemV design choices
+
+* AArch64 uses variant 1 TLS as described in ELFTLS_.
+
+* The thread pointer (TP) is always accessible via the ``TPIDR_EL0``
+  system register. This can be accessed via inlining a ``mrs``
+  instruction to read the thread pointer.
+
+* The compiler can generate code that supports a TLS block size of 4
+  KiB, 16 MiB, 4GiB or 16EiB, depending on the addressing mode. The
+  default is 16 MiB for all addressing modes.
+
+The static and dynamic linker must agree on the size of the padding
+between the TCB and the executables TLS Block. Using ``TCB`` as the
+size of the TCB (16 bytes), ``PAD`` as the size of the padding bytes,
+and ``PT_TLS`` as the program header with type PT_TLS. ``PAD`` must be
+the smallest positive integer that satisfies the following congruence:
+
+``TP + TCB + PAD ≡ PT_TLS.p_vaddr (modulo PT_TLS.p_align)``
```

Comment:
`TP+TCB+PAD` on the left could be confusing, as TCB is placed before TP. Perhaps mention the requirement of TP first (= 0 (modulo p_align)), then describe `PAD` and this formula.

Reply:
I'll see if I can word it better. I've found it difficult to try and explain the formula intuitively.